pkg/clients/clients.go

```diff
@@ -1,57 +1,163 @@
+
+const (
+	clusterVersionName = "version"
+)
+
 var (
-	log        = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
-	Interface  *ClientsInterface
-	RestConfig *rest.Config
-	Namespace  string
-	config     genericclioptions.ConfigFlags
+	log       = zap.New(zap.UseDevMode(true)).WithName(color.Print("clients", color.Brown))
+	Interface ClientsInterface
+	// TODO need to remove this global variable
+	Namespace string
 )
 
-type ClientsInterface struct {
+type ClientsInterface interface {
+	Update(ctx context.Context, obj client.Object) error
+	Get(ctx context.Context, key client.ObjectKey, obj client.Object) error
+	Delete(ctx context.Context, obj client.Object) error
+	List(ctx context.Context, obj client.ObjectList, opts ...client.ListOption) error
+	Create(ctx context.Context, obj client.Object) error
+	GetPodLogs(namespace, podName string, podLogOpts *v1.PodLogOptions) *restclient.Request
+	GetNamespace(ctx context.Context, name string, opts metav1.GetOptions) (*v1.Namespace, error)
+	GetSecret(ctx context.Context, namespace, name string, opts metav1.GetOptions) (*v1.Secret, error)
+	ClusterOperatorCreate(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.CreateOptions) (*configv1.ClusterOperator, error)
+	ClusterOperatorGet(ctx context.Context, name string, opts metav1.GetOptions) (*configv1.ClusterOperator, error)
+	ClusterOperatorUpdateStatus(ctx context.Context, clusterOperator *configv1.ClusterOperator, opts metav1.UpdateOptions) (*configv1.ClusterOperator, error)
+	ClusterVersionGet(ctx context.Context, opts metav1.GetOptions) (result *configv1.ClusterVersion, err error)
+	Invalidate()
+	ServerGroups() (*metav1.APIGroupList, error)
+	StatusUpdate(ctx context.Context, obj client.Object) error
+	CreateOrUpdate(ctx context.Context, obj client.Object, fn controllerutil.MutateFn) (controllerutil.OperationResult, error)
+	HasResource(resource schema.GroupVersionResource) (bool, error)
+	GetPlatform() (string, error)
+}
+
+type k8sClients struct {
 	client.Client
 	kubernetes.Clientset
 	clientconfigv1.ConfigV1Client
 	record.EventRecorder
 	authn.Keychain
 	discovery.CachedDiscoveryInterface
+	restConfig *restclient.Config
 }
 
-// GetKubeClientSet returns a native non-caching client for advanced CRUD operations
-func GetKubeClientSet() (*kubernetes.Clientset, error) {
-	return kubernetes.NewForConfig(RestConfig)
+func NewClients(runtimeClient client.Client, restConfig *restclient.Config, eventRecorder record.EventRecorder) (ClientsInterface, error) {
+	kubeClientSet, err := getKubeClientSet(restConfig)
+	if err != nil {
+		return &k8sClients{}, err
+	}
+	configClient, err := getConfigClient(restConfig)
+	if err != nil {
+		return &k8sClients{}, err
+	}
+	cachedDiscoveryClient, err := getCachedDiscoveryClient()
+	if err != nil {
+		return &k8sClients{}, err
+	}
+	return &k8sClients{
+		Client:                   runtimeClient,
+		Clientset:                *kubeClientSet,
```

Comment:
```suggestion
		Clientset:                *kubeClientSet,
```
ClientSet:                *kubeClientSet,

Reply:
mmm, not sure i understand what is the change

Reply:
Just typo - capital `S`

Reply:
right, saw it now. The code has already been changed for explicit fields, and not embedded ones

Reply:
fixed